File: Sources/DanmakuOverlay/Time/PlaybackClock.swift
import Foundation

/// 独立播放时钟：不依赖任何播放器，由用户手动校准（需求文档 7.4 节）
final class PlaybackClock {

    private(set) var isPlaying = false
    private var storedRate: Double = 1.0
    var rate: Double {
        get { storedRate }
        set {
            // 先按旧倍速结算已流逝的时间，再切换倍速，避免时间轴跳变。
            let settledTime = currentTime
            storedRate = min(max(newValue, 0.1), 4.0)
            baseTime = settledTime
            baseDate = now()
            onStateChange?()
        }
    }

    /// 时钟基准：currentTime = baseTime + (now - baseDate) * rate
    private var baseTime: Double = 0
    private var baseDate: Date
    private let now: () -> Date

    var onStateChange: (() -> Void)?

    var currentTime: Double {
        guard isPlaying else { return baseTime }
        return baseTime + now().timeIntervalSince(baseDate) * rate
    }

    init(now: @escaping () -> Date = Date.init) {
        self.now = now
        self.baseDate = now()
    }

    func play() {
        guard !isPlaying else { return }
        baseDate = now()
        isPlaying = true
        onStateChange?()
    }

    func pause() {
        guard isPlaying else { return }
        baseTime = currentTime
        isPlaying = false
        onStateChange?()
    }

    func toggle() {
        isPlaying ? pause() : play()
    }

    /// 跳转到绝对时间
    func seek(to time: Double) {
        baseTime = max(time, 0)
        baseDate = now()
        onStateChange?()
    }

    /// 相对偏移（正数快进，负数后退）
    func adjust(by delta: Double) {
        seek(to: currentTime + delta)
    }

    /// 将当前时刻设为 0 秒并开始播放（「从此刻同步」）
    func syncFromNow() {
        baseTime = 0
        baseDate = now()
        isPlaying = true
        onStateChange?()
    }
}

File: Windows/src/DanLaiDanQu.Core/PlaybackClock.cs
using System.Diagnostics;

namespace DanLaiDanQu.Core;

public sealed class PlaybackClock
{
    private readonly object _gate = new();
    private readonly Func<long> _timestamp;
    private readonly double _frequency;
    private double _baseTime;
    private long _baseTimestamp;
    private double _rate = 1.0;

    public PlaybackClock(Func<long>? timestamp = null, long? frequency = null)
    {
        _timestamp = timestamp ?? Stopwatch.GetTimestamp;
        _frequency = frequency ?? Stopwatch.Frequency;
        _baseTimestamp = _timestamp();
    }

    public event Action? Changed;

    public bool IsPlaying { get; private set; }

    public double Rate
    {
        get
        {
            lock (_gate)
            {
                return _rate;
            }
        }
        set
        {
            lock (_gate)
            {
                var now = _timestamp();
                _baseTime = CurrentTimeUnsafe(now);
                _baseTimestamp = now;
                _rate = Math.Clamp(value, 0.1, 4.0);
            }

            Changed?.Invoke();
        }
    }

    public double CurrentTime
    {
        get
        {
            lock (_gate)
            {
                return CurrentTimeUnsafe(_timestamp());
            }
        }
    }

    public void Play()
    {
        lock (_gate)
        {
            if (IsPlaying)
            {
                return;
            }

            _baseTimestamp = _timestamp();
            IsPlaying = true;
        }

        Changed?.Invoke();
    }

    public void Pause()
    {
        lock (_gate)
        {
            if (!IsPlaying)
            {
                return;
            }

            var now = _timestamp();
            _baseTime = CurrentTimeUnsafe(now);
            _baseTimestamp = now;
            IsPlaying = false;
        }

        Changed?.Invoke();
    }

    public void Toggle()
    {
        if (IsPlaying)
        {
            Pause();
        }
        else
        {
            Play();
        }
    }

    public void Seek(double time)
    {
        lock (_gate)
        {
            _baseTime = Math.Max(time, 0);
            _baseTimestamp = _timestamp();
        }

        Changed?.Invoke();
    }

    public void Adjust(double delta) => Seek(CurrentTime + delta);

    public void SyncFromNow()
    {
        lock (_gate)
        {
            _baseTime = 0;
            _baseTimestamp = _timestamp();
            IsPlaying = true;
        }

        Changed?.Invoke();
    }

    private double CurrentTimeUnsafe(long now)
    {
        if (!IsPlaying)
        {
            return _baseTime;
        }

        return _baseTime + ((now - _baseTimestamp) / _frequency * _rate);
    }
}
